### PR TITLE
Add kill by surronding

### PR DESCRIPTION
**Pull request by Simondharcourt**
Hello ! 
I saw your djambi project which I found amazing and wanted to contribute to the project.
Here is a suggestion to add the rule of the surronding of a chief.
Please let me know what you think of it.

### CODE REVIEW

**Comment by mabdelaal86** [comment]
Thanks a lot for the contribution. It is very appreciated.
I will review your PR as soon as possible.

File: flutter/lib/models/party.dart [hunk 1]
```diff
@@ -22,4 +21,6 @@ class Party {
 
   Member? getMemberAt(Cell cell) =>
       aliveMembers.firstWhereOrNull((m) => m.location == cell);
+
+  bool get isActive => chief.isAlive && !chief.isSurrounded() && movableMembers.isNotEmpty;
```
Comment (mabdelaal86) [inline flutter/lib/models/party.dart:25]: I prefer keeping `isActive` property close to `isLost`, as the logic is very related.

**Review by mabdelaal86 — CHANGES_REQUESTED** [review]
I will review the surrounding rule, and may come back with other comments.

File: flutter/lib/models/member.dart [hunk 2]
```diff
@@ -91,6 +92,45 @@ abstract class Member {
...
+    Cell(1, 1),   // top-right diagonal
+    Cell(-1, -1), // bottom-left diagonal
+    Cell(1, -1),  // bottom-right diagonal
+    Cell(-1, 1),  // top-left diagonal
+  ];
+
+  for (Cell direction in directions) {
+    Cell newPosition = Cell(location.x + direction.x, location.y + direction.y);
+    if (!newPosition.isValid) {
+      continue;  // Cell is off the board, considered as not surrounded.
```
[the remark this answers is not in the document]
Comment (Simondharcourt) [reply flutter/lib/models/member.dart]: Yes, indeed

**Comment by Simondharcourt** [comment]
> I will review the surrounding rule, and may come back with other comments.

Great, thanks for the review. I updated the code in consequence

**Comment by mabdelaal86** [comment]
I have reviewed the kill by surrounding rules, and played the game with your changes. I found many deviations between your implementation and the rules (based on my understanding).

The following is what stated regarding killing by surrounding in the document that you have added:

> The second great irony is that if a Leader becomes completely immobilized,
fully encircled by corpses either friend or foe that prevent him from moving, and does
not have a remaining Necromobile to rescue him by moving the surrounding dead to
a more favorable place of rest, he likewise perishes and is eliminated from the game.

> But the forces of a Leader who perishes by encirclement become satellites only
when a Leader first occupies the Labyrinth or already occupies it following the
encirclement. While waiting for devolution (transfer of control), the other players
have no ownership over the pieces of the encircled Leader.

So, I see the following deviations in your implementation:

1. If a Leader has a Necromobile, it is NOT considered encircled by corpses.
2. If a Leader encircled by corpses and doesn't have a Necromobile, he perishes (get killed).
3. Pieces of encircled Leader are controlled (or owned) by the Leader in the Maze (Labyrinth) or the first Leader occupies the Maze. 
4. While pieces are waiting for devolution (transfer of control), they can't be moved or killed (This part is from [Wikipedia](https://en.wikipedia.org/wiki/Djambi))

**Comment by Simondharcourt** [comment]
Thank you for your review. I will apply some changes to correct it as soon as possible.

**Comment by mabdelaal86** [comment]
I will close this PR since it was open for so long. I have implemented this feature in a slightly different way. You are welcome to contribute if you would like to make changes to the current behavior.

**Comment by Simondharcourt** [comment]
Sure ! I also started a repository some time ago to play a variant of the game, but I lack a bit of inspiration on how to adapt the rules..  if you want to have a look on it, you would be the most welcome :)) on https://github.com/Simondharcourt/djambi6players
(I made this project because I have the board game at home, and i feel the game needs a little bit of adaptation of the rules to be more balance and dynamic.. but I did not find the right balance yet)

**Comment by mabdelaal86** [comment]
Looks nice!

Since my objective is to play the game on mobile screens, I don't have plans to support more than 4 players in my project. I'm also trying to stick to Jean Anesto's original rules.

Feel free to get inspired by any part of my project, and let me know if you need any help in yours.

There is also another implementation of the game that supports 3 to 8 players:
https://github.com/JamesFaix/Djambi-N

**Pull request closed without merging**